src/slice/new_iter.rs

```diff
@@ -0,0 +1,495 @@
+//! Okay, this immutable iterator doesn't use any `Access` types, and thus could not be used
+//! on any element with possible mutable access at the same time (or could, but it's a subtle UB).
+//! It could be used without UB only with unaliased memory, and due to this, `.iter()` should produce
+//! a group of `(AliasedIter, UnAliasedIter, AliasedIter)` or use AliasedIter for every `.next()`.
+
+// This should be deleted when all unsafe blocks are used or deleted.
+#![allow(unused_unsafe)]
+
+#[cfg(feature = "std")]
+use std::{
+	iter::{
+		Chain, DoubleEndedIterator, ExactSizeIterator, FusedIterator, Iterator,
+	},
+	marker::PhantomData,
+	mem,
+	ptr::NonNull,
+};
+
+#[cfg(not(feature = "std"))]
+use core::{
+	iter::{
+		Chain, DoubleEndedIterator, ExactSizeIterator, FusedIterator, Iterator,
+	},
+	marker::PhantomData,
+	mem,
+	ptr::NonNull,
+};
+
+use crate::{
+	access::BitAccess,
+	index::{BitIdx, BitSel},
+	mem::BitMemory,
+	order::BitOrder,
+	pointer::BitPtr,
+	slice::BitSlice,
+	store::BitStore,
+};
+
+macro_rules! unchecked_add {
+	($x:expr, $y:expr) => {{
+		$x + $y
+		}};
+}
+
+macro_rules! unchecked_sub {
+	($x:expr, $y:expr) => {{
+		$x - $y
+		}};
+}
+
+/// Immutable bitslice iterator
+///
+/// This struct is created by the [`iter`] method on [bitslices].
+///
+/// [`iter`]: todo!()
+/// [bitslices]: todo!()
+pub struct Iter<'a, T: 'a + BitMemory, O> {
```

Comment:
The iterator layout is a complex problem. `bitvec` provides a lot of the facilities you remade in this PR, which given that you remade them rather than found them indicates that I have a **lot** of work to do on writing an onboarding guide for a new developer. That's my fault, not yours.

That said, `BitIdx` provides facilities for offset and span computations that we can leverage here. If we store two element addresses as `NonNull<T::Access>` (strictly speaking, only one needs to be `NonNull` while the other can be a bare pointer, but it's a harmless assertion to add) and two bit indices as `BitIdx<T::Mem>`, we can model a classic double-pointer iterator as a pair of address/bit selectors for the first live bit, and the first dead bit after the region.

> You're right that the use of `BitSel` over `BitIdx` allows for faster mask computation, by eliminating the `add` instruction and using only the appropriate shift instruction. I used to have `next` and `prev` methods in `BitOrder` to take a selector bit and compute the next/previous selector from it, but I removed them a while ago because they didn't generalize well and cause problems at boundary conditions. We can restore those pretty easily, but I'd rather prove correctness with the `BitIdx` model before trimming instructions.

And since `Iter` is semantically equivalent to a `&BitSlice`, we can just make a `PhantomData<&'a BitSlice<O, T>>` to satisfy the type checker.

Reply:
As you can see, I'm using rotations, not shifts. And while it would be hard to get same mechanics on shifts for non-standard orderings, rotation could be reproduced much easier.
And I haven't found any way to do `idx_from_sel`, so I made this one up. This one is even easier to make than rotation, as any ordering can simply make [(cttz(Sel) as u8, Idx), BITS] lookup table if there's no faster way.